docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+- [SnippetsTemplate](#snippetstemplate) which allows to bring unsupported NGINX configuration to NGF by Application
+  developers in a safe and uncomplicated manner. However, SnippetsTemplates require some up-front work from Cluster
+  operator, meaning it cannot be implemented quickly, in contrast with SnippetsPolicy.
+
+### SnippetsPolicy
+
+Snippets allow inserting NGINX configuration into various NGINX contexts.
+
+#### API
+
+> Field validation rules are intentionally left out of this proposal.
+
+```go
+// SnippetsPolicy is a Direct Attached Policy. It allows inserting NGINX configuration into the generated NGINX
+// config that affects resources referenced by the policy.
+type SnippetsPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the SnippetsPolicy.
+	Spec SnippetsPolicySpec `json:"spec"`
+
+	// Status defines the state of the SnippetsPolicy.
+	Status gatewayv1alpha2.PolicyStatus `json:"status,omitempty"`
+}
+
+// SnippetsPolicyList contains a list of SnippetsPolicies.
+type SnippetsPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []SnippetsPolicy `json:"items"`
+}
+
+// SnippetsPolicySpec defines the desired state of the SnippetsPolicy.
+type SnippetsPolicySpec struct {
+	// Snippets is a list of NGINX configuration snippets.
+	// There can only be one snippet per context.
+	Snippets []Snippet `json:"snippets"`
+
+	// TargetRefs identifies the API object(s) to apply the policy to.
+	// Objects must be in the same namespace as the policy.
+	// Objects must be of the same Kinds.
+	// TO-DO(pleshakov): Remove the line below.
+	// why same Kinds? Because different kinds allow mutually exclusive snippets.
+	//
+	// Support: Gateway, HTTPRoute, GRPCRoute, TLSRoute
+	//
+	// For HTTPRoute and GRPCRoute, only http* contexts are supported.
+	// For TLSRoute, only stream* contexts are supported.
+	TargetRefs []gatewayv1alpha2.LocalPolicyTargetReference `json:"targetRefs"`
+}
+
+// NginxContext represents the NGINX configuration context.
+type NginxContext string
+
+const (
+	// NginxContextMain is the main context of the NGINX configuration.
+	NginxContextMain NginxContext = "main"
+
+	// NginxContextHTTP is the http context of the NGINX configuration.
+	NginxContextHTTP NginxContext = "http"
+
+	// NginxContextHTTPServer is the server context of the NGINX configuration.
+	NginxContextHTTPServer NginxContext = "http.server"
+
+	// NginxContextHTTPServerLocation is the location context of the NGINX configuration.
+	// TO-DO(pleshakov): Decide if we need to explicitly support the parent location for rules with multiple matches.
+	// Same problem as in https://github.com/nginxinc/nginx-gateway-fabric/issues/2079
+	NginxContextHTTPServerLocation NginxContext = "http.server.location"
```

Comment:
Could we specify http.server.location when attaching to a Gateway? And then it would apply to all locations for that Gateway?

Reply:
yep,

I can convert this comment into a table or something like that or expand it:
```
	// For HTTPRoute and GRPCRoute, only http* contexts are supported.
	// For TLSRoute, only stream* contexts are supported.
```

as it was meant to cover that http.server.locations work for Gateways targets